### PR TITLE
Use std::unique_ptr<> for TensorFlow session

### PR DESCRIPTION
**Pull request by lissyx**

### CODE REVIEW

**Comment by lissyx** [comment]
```
diff --git a/native_client/client.cc b/native_client/client.cc
index 99af904e..92858a15 100644
--- a/native_client/client.cc
+++ b/native_client/client.cc
@@ -366,6 +366,19 @@ main(int argc, char **argv)
     return 1;
   }
 
+  for (int i = 0; i < 20; i++) {
+    ModelState* ctx;
+    int status = DS_CreateModel(model, beam_width, &ctx);
+    if (status != 0) {
+      fprintf(stderr, "Could not create model.\n");
+      return 1;
+    }
+    printf("DS_CreateModel/DS_FreeModel #%d\n", i);
+    DS_FreeModel(ctx);
+  }
+
+  return 0;
+
   // Initialise DeepSpeech
   ModelState* ctx;
   int status = DS_CreateModel(model, beam_width, &ctx);
```

```
$ tail -n 12 std-unique_ptr.*
==> std-unique_ptr.0.log <==
==25363== 
==25363== LEAK SUMMARY:
==25363==    definitely lost: 32 bytes in 1 blocks
==25363==    indirectly lost: 72 bytes in 3 blocks
==25363==      possibly lost: 330,996 bytes in 1,472 blocks
==25363==    still reachable: 1,551,566 bytes in 33,791 blocks
==25363==                       of which reachable via heuristic:
==25363==                         stdstring          : 396,638 bytes in 11,288 blocks
==25363==                         newarray           : 3,792 bytes in 9 blocks
==25363==         suppressed: 0 bytes in 0 blocks
==25363== 
==25363== ERROR SUMMARY: 28 errors from 28 contexts (suppressed: 0 from 0)

==> std-unique_ptr.1.log <==
==29407== 
==29407== LEAK SUMMARY:
==29407==    definitely lost: 29,920 bytes in 25 blocks
==29407==    indirectly lost: 6,108,730 bytes in 81,381 blocks
==29407==      possibly lost: 610,067 bytes in 5,140 blocks
==29407==    still reachable: 2,009,189 bytes in 39,701 blocks
==29407==                       of which reachable via heuristic:
==29407==                         stdstring          : 528,199 bytes in 12,793 blocks
==29407==                         newarray           : 30,320 bytes in 134 blocks
==29407==         suppressed: 0 bytes in 0 blocks
==29407== 
==29407== ERROR SUMMARY: 321 errors from 321 contexts (suppressed: 0 from 0)
```

Where `std-unique_ptr.0.log` includes the move to `std::unique_ptr<>` and `std-unique_ptr.1.log` is with plain pointers.

File: native_client/tfmodelstate.cc [hunk 1]
```diff
@@ -50,17 +49,19 @@ TFModelState::init(const char* model_path,
     options.config.mutable_graph_options()
       ->mutable_optimizer_options()
       ->set_opt_level(::OptimizerOptions::L0);
-    options.env = mmap_env_;
+    options.env = mmap_env_.get();
   }
 
-  status = NewSession(options, &session_);
+  Session* _session;
```
Comment (reuben) [inline native_client/tfmodelstate.cc]: `session`, not `_session`.

File: native_client/tfmodelstate.cc [hunk 2]
```diff
@@ -50,17 +49,19 @@ TFModelState::init(const char* model_path,
...
+  status = NewSession(options, &_session);
   if (!status.ok()) {
     std::cerr << status << std::endl;
     return DS_ERR_FAIL_INIT_SESS;
   }
+  session_ = std::unique_ptr<tensorflow::Session>(_session);
```
Comment (reuben) [inline native_client/tfmodelstate.cc]: `session_.reset(session);`

File: native_client/tfmodelstate.cc (hunk 1 above)
Comment (lissyx) [reply native_client/tfmodelstate.cc]: Done

File: native_client/tfmodelstate.cc (hunk 2 above)
Comment (lissyx) [reply native_client/tfmodelstate.cc]: Always forget about `.reset()`

Comment (lissyx) [reply native_client/tfmodelstate.cc]: We could also `.reset()` for `mmap_env_` :)

**Pull request merged**